File: cpp/implement_queue_using_stacks.cpp
232. Implement Queue using Stacks  
Implement the following operations of a queue using stacks.

push(x) -- Push element x to the back of queue.
pop() -- Removes the element from in front of queue.
peek() -- Get the front element.
empty() -- Return whether the queue is empty.
Notes:
You must use only standard operations of a stack -- which means only push to top, peek/pop from top, size, and is empty operations are valid.
Depending on your language, stack may not be supported natively. You may simulate a stack by using a list or deque (double-ended queue), as long as you use only standard operations of a stack.
You may assume that all operations are valid (for example, no pop or peek operations will be called on an empty queue).
Hide Company Tags Microsoft Bloomberg
Hide Tags Stack Design
Hide Similar Problems (E) Implement Stack using Queues

class MyQueue {
public:
    /** Initialize your data structure here. */
    MyQueue() {
        
    }
    
    /** Push element x to the back of queue. */
    void push(int x) {
        if (in.empty()) front=x;
        in.push(x);
    }
    
    /** Removes the element from in front of queue and returns that element. */
    int pop() {
        if (empty()) return -1;
        if (out.empty()) {
            while (!in.empty()) {
                out.push(in.top());
                in.pop();
            }
        }
        int val=out.top();
        out.pop();
        return val;
    }
    
    /** Get the front element. */
    int peek() {
        if (empty()) return -1;
        if (out.empty()) return front;
        else return out.top();
    }
    
    /** Returns whether the queue is empty. */
    bool empty() {
        return in.empty() && out.empty();
    }
private:
    stack<int> in;
    stack<int> out;
    int front;
};

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue obj = new MyQueue();
 * obj.push(x);
 * int param_2 = obj.pop();
 * int param_3 = obj.peek();
 * bool param_4 = obj.empty();
 */

File: python/implement-queue-using-stacks.py
class MyQueue:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.int=[]
        self.out=[]
        self.front=0

    def push(self, x: int) -> None:
        """
        Push element x to the back of queue.
        """
        if len(self.int)==0: self.front=x
        self.int.append(x)

    def pop(self) -> int:
        """
        Removes the element from in front of queue and returns that element.
        """
        if self.empty(): return
        if len(self.out)==0:
            while self.int:
                self.out.append(self.int.pop())
        return self.out.pop()
            

    def peek(self) -> int:
        """
        Get the front element.
        """
        if self.empty(): return -1
        if len(self.int)>0: return self.front
        else: return self.out[-1]        

    def empty(self) -> bool:
        """
        Returns whether the queue is empty.
        """
        return len(self.int)==0 and len(self.out)==0


# Your MyQueue object will be instantiated and called as such:
# obj = MyQueue()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.peek()
# param_4 = obj.empty()

